File: src/commands/pull.ts
import { getAccountByName, getTransactionsByAccountName } from '../services/plaid';
import { getAccessToken, getCachedTransactions, hasAccessToken } from '../utils/storage';

interface PullOptions {
  month: string;
  format: string;
}

export async function pullStatements(options: PullOptions) {
  console.log(`Pulling statements for ${options.month} in ${options.format} format...`);
  
  if (!hasAccessToken()) {
    console.error('❌ No linked account found. Please run: npm run dev link');
    console.log('Then complete the linking process and exchange your public token.');
    return;
  }

  // TODO: check cache conditional here
  // const cachedTransactions = getCachedTransactions(options.month);
  // if (!cachedTransactions) {
  //   console.error('❌ No cached transactions found for the specified month.');
  //   return;
  // }

  const accessToken = getAccessToken()!;

  try {
    const [year, month] = options.month.split('-');
    const startDate = `${year}-${month.padStart(2, '0')}-01`;
    const endOfMonth = new Date(parseInt(year), parseInt(month), 0);
    const endDate = `${year}-${month.padStart(2, '0')}-${endOfMonth.getDate().toString().padStart(2, '0')}`;

    console.log(`📊 Fetching transactions from ${startDate} to ${endDate}...`);
    
    const [acc, transactions] = await Promise.all([
      getAccountByName('RBC ION+ Visa', accessToken),
      getTransactionsByAccountName('RBC ION+ Visa', accessToken, startDate, endDate)
    ]);

    const totalExpenses = transactions
      .filter(tx => tx.amount > 0)
      .reduce((sum, tx) => sum + tx.amount, 0);

    const data = {
      month: options.month,
      accounts: {
        name: acc.name,
        type: acc.type,
        balance: acc.balance.current
      },
      transactions: transactions.map(tx => ({
        date: tx.date,
        description: tx.name,
        amount: -tx.amount, // Flip sign for intuitive display
        category: tx.category
      })),
      summary: {
        totalExpenses: Math.round(totalExpenses * 100) / 100,
        transactionCount: transactions.length
      }
    };

    if (options.format === 'json') {
      console.log(JSON.stringify(data, null, 2));
    } else if (options.format === 'csv') {
      console.log('Date,Description,Amount,Category');
      data.transactions.forEach(tx => {
        console.log(`${tx.date},"${tx.description}",${tx.amount},${tx.category}`);
      });
    }

    console.log(`\n✅ Found ${transactions.length} transactions for ${options.month}`);
    
  } catch (error) {
    console.error('❌ Failed to pull statements:', error);
    console.log('💡 Make sure your account is still linked. You may need to re-link if the token expired.');
  }
}

File: src/budget_tracker/commands/pull.py
"""Pull command implementation."""

import os
from typing import Optional

from budget_tracker.services.plaid_client import PlaidClient
from budget_tracker.utils.storage import (
    get_access_token,
    get_cached_transactions,
    save_cached_transactions,
    save_cached_transactions_csv,
)


def pull_statements(
    year: Optional[str] = None, month: Optional[str] = None, format: str = "json"
) -> None:
    """Pull monthly bank statements."""
    print(f"Month: {month or 'current'}, Format: {format}")

    plaid_client = PlaidClient()
    access_token = get_access_token()
    accounts = plaid_client.get_accounts(access_token)

    accountFindByName = next(
        (
            account
            for account in accounts
            if account["name"] == os.getenv("ACCOUNT_TO_FILTER")
        ),
        None,
    )

    if not accountFindByName:
        print(
            "❌ Account not found! Check your ACCOUNT_TO_FILTER environment variable."
        )
        return

    transactions = get_cached_transactions(accountFindByName["name"], year, month)
    if transactions:
        print(f"✅ Found cached transactions: {len(transactions)} transactions")
    else:
        print("❌ No cached transactions found")
        transactions = plaid_client.get_transactions(
            access_token, accountFindByName["account_id"], year, month
        )
        transactions = [tx.to_dict() for tx in transactions]
        save_cached_transactions(accountFindByName["name"], year, month, transactions)

        transactionsPartialData = []
        for tx in transactions:
            if tx and tx.get("amount", 0) > 0:
                pfc = tx.get("personal_finance_category") or {}
                transactionsPartialData.append(
                    {
                        "date": tx["date"],
                        "description": tx["name"],
                        "amount": tx["amount"],
                        "category": pfc.get("primary", "Unknown"),
                        "subcategory": pfc.get("detailed", "Unknown"),
                        "category_confidence": pfc.get("confidence_level", "Unknown"),
                    }
                )

        categories = set(tx["category"] for tx in transactionsPartialData)
        summaryData = {
            "totalSpending": sum(
                transaction["amount"]
                for transaction in transactions
                if transaction and transaction.get("amount", 0) > 0
            ),
            "transactionCount": len(transactions),
            "totalSpendingByCategory": {
                category: sum(
                    transaction["amount"]
                    for transaction in transactions
                    if transaction
                    and transaction.get("amount", 0) > 0
                    and (transaction.get("personal_finance_category") or {}).get(
                        "primary"
                    )
                    == category
                )
                for category in categories
            },
            "totalTransactionsCountByCategory": {
                category: sum(
                    1
                    for transaction in transactions
                    if transaction
                    and transaction.get("amount", 0) > 0
                    and (transaction.get("personal_finance_category") or {}).get(
                        "primary"
                    )
                    == category
                )
                for category in categories
            },
        }

        save_cached_transactions_csv(
            accountFindByName["name"], year, month, transactionsPartialData, summaryData
        )

    totalSpending = sum(
        transaction["amount"]
        for transaction in transactions
        if transaction["amount"] > 0
    )
    print(f"Total spending: {totalSpending}")
